src/iconforge.rs

```diff
@@ -0,0 +1,1088 @@
+            .into_par_iter()
+            .for_each(|(transform, mut associated_icons)| {
+                let mut altered_image;
+                {
+                    zone!("clone_image");
+                    altered_image = image.clone();
+                }
+                if let Err(err) = transform_image(&mut altered_image, &transform) {
+                    errors.lock().unwrap().push(err);
+                }
+                {
+                    zone!("filter_associated_icons");
+                    associated_icons
+                        .clone()
+                        .into_iter()
+                        .enumerate()
+                        .for_each(|(idx, icon)| {
+                            if icon.transform.len() as u8 == depth + 1
+                                && *icon.transform.last().unwrap() == transform
+                            {
+                                associated_icons.swap_remove(idx);
+                                if let Err(err) = return_image(altered_image.clone(), icon) {
+                                    errors.lock().unwrap().push(err.to_string());
+                                }
+                            }
+                        });
+                }
+                if let Err(err) = transform_leaves(&associated_icons, altered_image, depth + 1) {
+                    errors.lock().unwrap().push(err);
+                }
+            });
+    }
+
+    if !errors.lock().unwrap().is_empty() {
+        return Err(errors.lock().unwrap().join("\n"));
+    }
+    Ok(())
+}
+
+/// Converts an IO icon to one with icon_hash_input
+fn icon_from_io(icon_in: IconObjectIO) -> IconObject {
+    zone!("icon_from_io");
+    // TODO: can probably convert this function to just lazily attaching icostring to a RefCell<> or something
+    // This alternative type system is too verbose and wasteful of processing time.
+    // https://doc.rust-lang.org/reference/interior-mutability.html
+    let mut result = IconObject {
+        icon_file: icon_in.icon_file,
+        icon_state: icon_in.icon_state,
+        dir: icon_in.dir,
+        frame: icon_in.frame,
+        transform: icon_in
+            .transform
+            .into_iter()
+            .map(|transform_in| match transform_in {
+                TransformIO::BlendColor { color, blend_mode } => {
+                    Transform::BlendColor { color, blend_mode }
+                }
+                TransformIO::BlendIcon { icon, blend_mode } => Transform::BlendIcon {
+                    icon: icon_from_io(icon),
+                    blend_mode,
+                },
+                TransformIO::Crop { x1, y1, x2, y2 } => Transform::Crop { x1, y1, x2, y2 },
+                TransformIO::Scale { width, height } => Transform::Scale { width, height },
+            })
+            .collect(),
+        transform_hash_input: String::new(),
+        icon_hash_input: String::new(),
+    };
+    result.gen_icon_hash_input().unwrap(); // unsafe but idc
```

Comment:
?

Reply:
The unwrap() could technically panic. I can add a fail check.